app/models/metrics/abstract_metrics_factory.py

```diff
@@ -0,0 +1,116 @@
+
+        :param module_id: The first part of the primary key
+        :param ids: Used in the second part of the primary key.
+        :return: dictionary of ClickdataModel objects keyed on item (i.e. not including the prefix)
+        """
+        # Keys are namespaced by the module we are getting data from
+        keys = {self._make_key(module_id, i) for i in ids}
+
+        # Always get the default key, it is used for items that don't have any metrics
+        keys.add(self._make_key(module_id, "default"))
+        keys = list(keys)
+
+        metrics = await self._query_cached_metrics(keys)
+        # Remove "/<modules>" suffix and remove None values
+        # TODO: It might be cleaner if this method just returns List[MetricsBaseModel], and callers create the dict
+        # of their choosing.
+        metrics = {k.split("/")[0]: self.parse_from_record(v) for k, v in metrics.items() if v is not None}
+
+        if not metrics:
+            logging.error(f"No metrics for module {module_id} with keys={keys}")
+
+        return metrics
+
+    def parse_from_record(self, value: Dict):
+        """
+        Similar to Pydantic's parse_obj function, but sets the id field using the primary key.
+        :param value: dictionary containing all required keys for MetricsBaseModel
+        :return: Parsed MetricsBaseModel
+        """
+        return MetricsModel.parse_obj({**value, 'id': value[self._primary_key_name]})
+
+    @xray_recorder.capture_async('models.metrics.MetricsBaseModel._query_cached_metrics')
+    async def _query_cached_metrics(self, metrics_keys: List) -> Dict[str, Optional[Dict]]:
+        multi_cache_wrapper = decorators.multi_cached(
+            keys_from_attr="metrics_keys",
+            ttl=app.config.elasticache['metrics_ttl'],
+            alias=app.cache.metrics_alias)
+
+        multi_cache = multi_cache_wrapper(self._query_metrics)
+
+        # multi_cache will get metrics from cache when available, and call _query_metrics with all missing keys.
+        results = await multi_cache(metrics_keys=metrics_keys)
+
+        # Map app.cache's NoneValue to None. aiocache treats None as a cache miss, so we need a special token for this.
+        return {k: None if v == app.cache.NoneValue else v for k, v in results.items()}
+
+    @xray_recorder.capture_async('models.MetricsBaseModel._query_metrics')
+    async def _query_metrics(self, metrics_keys: List):
+        metrics = {}
+
+        async with aioboto3.resource('dynamodb', endpoint_url=self._dynamodb_endpoint) as dynamodb:
+            for keychunk in _chunks(metrics_keys):
+                request = {
+                    self._dynamodb_table: {
+                        "Keys": [{self._primary_key_name: c} for c in keychunk]
+                    }
+                }
+                responses = await dynamodb.batch_get_item(RequestItems=request)
+
+                # TODO: Write a unit test when there are more than 100  rows. It seems this will fail because it's only
```

Comment:
TODO: Write unit test for case where there are more than 100 rows

Reply:
This was a preexisting bug, unrelated to this PR, that I fixed here. It would be good to write a unit test to prevent it from regressing.